README.md
# Haskell-Webservice

## Notes and Instuctions

** Please read Jim's note in oldREADME.md before this if you are new

This project contains a web service written in Haskell using Happstack and SQLite.
The web service provides an interface to (fictitious, randomly generated) weather 
readings taken at the North Pole during 2017.

** Home Page (GET, POST): http://localhost:8000/weather/
Show some messages and instructions.

** Single date query (GET): weather/date/YYYY-mm-dd
Return JSON file if weather data for that date is held.

** Insert/Update date query (PUT):
Insert or update a date/temperature pair in the database.
*  webservice:

    $ curl -X PUT weather/put/YYYY-mm-dd/temperature

*  oldwebservice:

    $ curl -X PUT weather/date/YYYY-mm-dd/temperature

** Range dates query (GET): weather/range/d1/d2
**                   where: d1  = YYYY-mm-dd
**                          d2  = YYYY-mm-dd
**                          d1  != d2
Return JSON file if weather data for dates between date d1 and d2 are held.

** Max dates query (GET): weather/max/d1/d2
**                 where: d1  = YYYY-mm-dd
**                        d2  = YYYY-mm-dd
**                        d1  != d2
Return JSON file if weather data for dates between date d1 and d2 with highest temperature are held.

** Above temperature query (GET): weather/above/t
**                         where: t is a real number
Return JSON file if weather data for dates that have temperature higher than t.

** Return 405 Method not allowed and an empty JSON object on wrong request methods.

## Installation

** If you are using Windows, you should install CygWin and use the CygWin terminal instead
**of the Windows command prompt, as several of the packages you need require a bash-like environment.

** If you are working on this code in the labs, do not install it on a Windows drive
**(e.g. `~/W_DRIVE` or `~/M_DRIVE`), as cabal is known to have problems with Windows network shares.

** Check for update:

    $ cabal update
    $ cabal install cabal-install
	
** Clone project

    $ git clone https://github.com/ChrisTran96/Haskell-Webservice.git
    $ cd haskell-webservice

** Create sandbox (optional)

    $ cabal sandbox init

** Configure and install missing package

    $ cabal configure

** Build and install

	  $ cabal build
	  $ cabal install

** Run DB-setup

	  $ cabal run DB-setup

**Running the weather service
*Run webservice

    $ cabal run webservice

*Run old webservice

    $ cabal run oldwebservice

**Try test (work on both webservice and oldwebservice)
While THE SERIVICE is RUNNING open a new terminal and run the tests:

    $ cabal test

A log can be found at:
dist/test/haskell-webservice-0.1.0.0-test-webservice.log



ChangeLog.md
# Revision history for haskell-webservice

## 0.1.0.0  -- YYYY-mm-dd

* First version. Released on an unsuspecting world.

## 0.1.5.0  -- 2018-04-24

* Implement range/d1/d2
* Implement max/d1/d2
* Implement above/temperature
* Response 405 "Method not allowed" on illigel request (Not fully working yet)

## 0.2.0.0  -- 2018-04-26

* Implement Web Routes library

## 0.2.5.0  -- 2018-05-01

* Adding method into WebRoutes version
* Update cabal file
* Fix the "response 405"
* Improve code style

tests/Main.hs
{-# LANGUAGE OverloadedStrings #-}
module Main where

import Test.Tasty (testGroup)
import Test.HTTP
import Data.List (isInfixOf)
import Data.Aeson (decode)
import WeatherService.OldService
import qualified Data.ByteString.Lazy.Char8 as LB

url = "http://localhost:8000/weather/"

day0 = "{\"date\":\"2017-01-01\",\"temperature\":-23.164497}"
hottestDay = "[{\"date\":\"2017-09-15\",\"temperature\":39.502533}]"

testDayQuery = httpTestCase "Weather Service Day Query" url $ do
    landing <- get "date/2017-01-01"
    assert "Includes initial data" $ day0 `isInfixOf` landing
    debug landing

testRangeQuery = httpTestCase "Weather Service Range Query" url $ do
    landing <- get "range/2017-01-01/2017-12-31" --should be 365 records
    let days = decode (LB.pack landing) :: Maybe [WeatherField]
    case days of
      Nothing   -> assert "Has year of records" $ False
      (Just ms) -> assert "Has year of records" $ length ms == 365

testMaxQuery = httpTestCase "Weather Service Max Query" url $ do
    landing <- get "max/2017-01-01/2017-12-31"
    assert "Has max record" $ landing == hottestDay

testAboveQuery = httpTestCase "Weather Service Above Query" url $ do
    landing <- get "above/39" --should be two records
    let days = decode (LB.pack landing) :: Maybe [WeatherField]
    debug (show days)
    case days of
      Nothing   -> assert "Has Above record" $ False
      (Just ms) -> assert "Has Above record" $ length ms == 2

tests = testGroup "tests" [testDayQuery, testRangeQuery, testMaxQuery, testAboveQuery]

main = defaultMain tests


src/Main.hs
{-# LANGUAGE OverloadedStrings, FlexibleContexts, TypeFamilies #-}
module Main where
{-| Semester 2 assignment for CI285, University of Brighton
    Jim Burton <j.burton@brighton.ac.uk>
    Modifier: Chris Tran - 15800120
-}
import System.Log.Logger      ( updateGlobalLogger
                              , rootLoggerName
                              , setLevel
                              , debugM
                              , Priority(..)
                              )
import Control.Monad          (msum)
import Data.Text              (pack, unpack)

import Happstack.Server
import Database.SQLite.Simple
import Web.Routes.Happstack   (implSite)
import qualified Data.ByteString.Lazy.Char8 as BC

import WeatherService.Service

{-| Entry point -}
main :: IO()
main = do
  updateGlobalLogger rootLoggerName (setLevel INFO) -- change level to DEBUG for testing
  conn <- open "data/np-weather.db"
  simpleHTTP nullConf $ do
    setHeaderM "Content-Type" "application/json"
    msum [
      implSite "http://localhost:8000" "/weather" (site conn)
      ]


src/WRMain.hs
{-# LANGUAGE OverloadedStrings, FlexibleContexts, TypeFamilies #-}
module Main where
{-| Semester 2 assignment for CI285, University of Brighton
    Web-Routes
    Author: Chris Tran - 15800120
-}
import System.Log.Logger      ( updateGlobalLogger
                              , rootLoggerName
                              , setLevel
                              , debugM
                              , Priority(..)
                              )
import Control.Monad          (msum)
import Data.Text              (pack, unpack)

import Happstack.Server
import Database.SQLite.Simple
import Web.Routes.Happstack   (implSite)
import qualified Data.ByteString.Lazy.Char8 as BC

import WeatherService.WRService

{-| Entry point -}
main :: IO()
main = do
  updateGlobalLogger rootLoggerName (setLevel INFO) -- change level to DEBUG for testing
  conn <- open "data/np-weather.db"
  simpleHTTP nullConf $ msum [
    implSite "http://localhost:8000" "/weather" (site conn)
    --Currently, this message can be seen when typing wrong url address
    , resp 405 $ toResponse $ BC.pack "Error 405\n\
                                      \Method is not allowed!\n\
                                      \Please go back to Home page:\n\
                                      \http://localhost:8000/weather\n"
    ]

src/OldMain.hs
{-# LANGUAGE OverloadedStrings, FlexibleContexts, TypeFamilies #-}
module Main where
{-| Semester 2 assignment for CI285, University of Brighton
    Jim Burton <j.burton@brighton.ac.uk>
    Modifier: Chris Tran - 15800120
-}
import           System.Log.Logger ( updateGlobalLogger
                                   , rootLoggerName
                                   , setLevel
                                   , debugM
                                   , Priority(..)
                                   )
import           Control.Monad          (msum)
import           Happstack.Server  
import           Database.SQLite.Simple
import           Data.Text              (pack)

import           Control.Applicative

import           WeatherService.OldService

{-| Entry point. Connects to the database and passes the connection to the
routing function. -}
main :: IO()
main = do
  updateGlobalLogger rootLoggerName (setLevel INFO) -- change level to DEBUG for testing
  conn <- open "data/np-weather.db"
  simpleHTTP nullConf $ do
    setHeaderM "Content-Type" "application/json"
    msum [
      dirs "weather/date" $
        msum [
        do method [GET, POST]
           path $ \d -> dayHandler d conn
        , do method PUT
             path $ \d -> path $ \t -> dayPutHandler d t conn
        , methodNotAllowedHandler
        ]
      {-, dirs "weather/date" $ do method PUT
                                 path $ \d -> path $ \t -> dayPutHandler d t conn-}
      , dirs "weather/range" $
        msum [
          do method GET
             path $ \d1 -> path $ \d2 -> rangeHandler d1 d2 conn
          , methodNotAllowedHandler
          ]
      , dirs "weather/max" $
        msum [
          do method GET
             path $ \d1 -> path $ \d2 -> maxHandler d1 d2 conn
          , methodNotAllowedHandler
          ]
      , dirs "weather/above" $
        msum [
          do method GET
             path $ \t -> aboveTHandler t conn
          , methodNotAllowedHandler
          ]
      ]


src/WeatherService/WRService.hs
{-# LANGUAGE OverloadedStrings, FlexibleContexts, TypeFamilies
  , QuasiQuotes, TemplateHaskell, DeriveGeneric #-}
module WeatherService.WRService (WeatherField(..)
                                , site
                                , homePage
                                , dayHandler
                                , putHandler
                                , rangeHandler
                                , maxHandler
                                , aboveTHandler) where
{-| Semester 2 assignment for CI285, University of Brighton
    Jim Burton <j.burton@brighton.ac.uk>
    Modifier: Chris Tran - 15800120
    This file was done base on Service.hs
-}
import           System.Log.Logger          (debugM)
import           Control.Monad.IO.Class     (liftIO)
import           Control.Monad              (msum)
import           Data.List                  (intercalate)
import           Data.Text                  (Text, pack, unpack)

import           Happstack.Server
import           Web.Routes
import           Web.Routes.Happstack
import           Web.Routes.TH

import           Database.SQLite.Simple
import           Database.SQLite.Simple.FromRow

import           Data.Aeson
import           GHC.Generics               (Generic)
import qualified Data.ByteString.Lazy.Char8 as BC

data WeatherField = WeatherField {date :: Text, temperature :: Float}
                    deriving (Generic, Show)

instance FromRow WeatherField where -- ^ Marshal data from DB to our ADT
  fromRow = WeatherField <$> field <*> field

instance ToRow WeatherField where -- ^ Marshal data from our ADT to the DB
  toRow (WeatherField theDate temp) = toRow (theDate, temp)

instance ToJSON   WeatherField -- ^ Marshal data from our ADT to JSON
instance FromJSON WeatherField -- ^ Marshal data from JSON to our ADT

{-| Site Map -}
data Sitemap
  = Weather           --weather/                      (home page)
  | Date Text         --weather/date/YYYY-mm-dd
  | Put Text Text     --weather/put/date/temperature
  | Range Text Text   --weather/range/date1/date2
  | Max Text Text     --weather/max/date1/date2
  | Above Text        --weather/above/temperature

$(derivePathInfo ''Sitemap)

{-| Site Route -}
route :: Connection -> Sitemap -> RouteT Sitemap (ServerPartT IO) Response
route conn url =
  case url of
    Weather        -> do method [GET, POST]
                         homePage
    (Date d)       -> do method [GET, POST]
                         dayHandler d conn
    (Put d t)      -> do method PUT
                         putHandler d t conn
    (Range d1 d2)  -> do method GET
                         rangeHandler d1 d2 conn
    (Max d1 d2)    -> do method GET
                         maxHandler d1 d2 conn
    (Above t)      -> do method GET
                         aboveTHandler t conn

{-| Home Page -}
homePage :: RouteT Sitemap (ServerPartT IO) Response
homePage = ok $ toResponse $
              BC.pack "Home Page\n\
                      \Haskell Weather Service by Chris Tran\n\
                      \\n\
                      \*Note: d = date (YYYY-mm-dd)\n\
                      \       t = temperature\n\
                      \\n\
                      \To check temperature of a single day:\n\
                      \  /weather/date/d\n\
                      \\n\
                      \To insert new day or update new temperature:\n\
                      \In terminal:\n\
                      \  $ curl -X PUT http://localhost:8000/weather/put/d/t\n\
                      \\n\
                      \To check temperature in between a range:\n\
                      \  /weather/range/d1/d2\n\
                      \\n\
                      \To find day(s) that has max temperature in between a range:\n\
                      \  /weather/max/d1/d2\n\
                      \\n\
                      \To find day(s) that has higher temperature than t:\n\
                      \  /weather/date/d\n"

{-| Handle requests for a single date. -}
dayHandler :: Text -> Connection -> RouteT Sitemap (ServerPartT IO) Response
dayHandler d conn = do
  setHeaderM "Content-Type" "application/json"
  r <- liftIO (queryNamed conn "SELECT the_date, temperature \
                               \ FROM  weather \
                               \ WHERE the_date = :dt" [":dt" := d] :: IO [WeatherField])
  liftIO $ debugM "Date Query" $ listToOutput r
  case r of
    [] -> notFoundHandler
    _  -> ok $ toResponse $ listToOutput r

{-| Handle PUT requests for date/temperature pairs. -}
putHandler :: Text -> Text -> Connection -> RouteT Sitemap (ServerPartT IO) Response
putHandler d t conn = do
  r <- liftIO (queryNamed conn "SELECT the_date, temperature \
                               \ FROM  weather \
                               \ WHERE the_date = :dt" [":dt" := d] :: IO [WeatherField])
  liftIO $ debugM "Put request" $ listToOutput r
  case r of
    [] -> insertHandler d t conn
    _  -> updateHandler d t conn

{-| Insert a new date/temperature pair. -}
insertHandler :: Text -> Text -> Connection -> RouteT Sitemap (ServerPartT IO) Response
insertHandler d t conn = do
  let t' = (read $ unpack t)::Float
  liftIO (execute conn "INSERT INTO weather (the_date, temperature) \
                       \ VALUES (?,?)" (WeatherField d t'))
  ok emptyJSONResponse

{-| Update a date/temperature pair. -}
updateHandler :: Text -> Text -> Connection -> RouteT Sitemap (ServerPartT IO) Response
updateHandler d t conn = do
  liftIO (executeNamed conn "UPDATE weather \
                          \ SET temperature = :dt1 \
                          \ WHERE the_date = :dt2"
                          [":dt1" := t, ":dt2" := d])
  ok emptyJSONResponse

{-| Return 404 Not Found and an empty JSON object -}
notFoundHandler :: RouteT Sitemap (ServerPartT IO) Response
notFoundHandler = notFound emptyJSONResponse

{-| An empty JSON object -}
emptyJSONResponse = toResponse $ BC.pack "[]"

{-| Turn a list of WeatherFields into a JSON object. -}
listToOutput :: ToJSON a => [a] -> String
listToOutput xs = "[" ++ intercalate "," (map (BC.unpack . encode) xs) ++ "]"

{-| Handle requests for dates in between date 1 and date 2.-}
rangeHandler :: Text -> Text -> Connection -> RouteT Sitemap (ServerPartT IO) Response
rangeHandler d1 d2 conn = do
  setHeaderM "Content-Type" "application/json"
  r <- liftIO (queryNamed conn "SELECT the_date, temperature \
                               \ FROM  weather \
                               \ WHERE the_date BETWEEN :dt1 AND :dt2"
                               [":dt1" := d1,":dt2" := d2] :: IO [WeatherField])
  liftIO $ debugM "Range Query" $ listToOutput r
  case r of
    [] -> notFoundHandler
    _  -> ok $ toResponse $ listToOutput r

{-| Handle requests for dates in between date 1 and date 2 with max temperature.-}
maxHandler :: Text -> Text -> Connection -> RouteT Sitemap (ServerPartT IO) Response
maxHandler d1 d2 conn = do
  setHeaderM "Content-Type" "application/json"
  r <- liftIO (queryNamed conn "SELECT the_date, MAX(temperature) \
                               \ FROM  weather \
                               \ WHERE the_date BETWEEN :dt1 AND :dt2"
                               [":dt1" := d1,":dt2" := d2] :: IO [WeatherField])
  liftIO $ debugM "Max Query" $ listToOutput r
  case r of
    [] -> notFoundHandler
    _  -> ok $ toResponse $ listToOutput r

{-| Handle requests for dates that has a higher temperature than t. -}
aboveTHandler :: Text -> Connection -> RouteT Sitemap (ServerPartT IO) Response
aboveTHandler t conn = do
  setHeaderM "Content-Type" "application/json"
  r <- liftIO (queryNamed conn "SELECT the_date, temperature \
                               \ FROM  weather \
                               \ WHERE temperature >= :dt"
                               [":dt" := t] :: IO [WeatherField])
  liftIO $ debugM "Above Query" $ listToOutput r
  case r of
    [] -> notFoundHandler
    _  -> ok $ toResponse $ listToOutput r

{-| Site with Weather set as Home Page -}
site :: Connection -> Site Sitemap (ServerPartT IO Response)
site conn =
            setDefault Weather $ mkSitePI (runRouteT (route conn))

src/WeatherService/OldService.hs
{-# LANGUAGE OverloadedStrings, FlexibleContexts, TypeFamilies,
    DeriveGeneric #-}
module WeatherService.OldService (WeatherField(..)
                              , dayHandler
                              , dayPutHandler
                              , rangeHandler
                              , maxHandler
                              , aboveTHandler
                              , methodNotAllowedHandler) where
{-| Semester 2 assignment for CI285, University of Brighton
    Jim Burton <j.burton@brighton.ac.uk>
    Modifier: Chris Tran - 15800120
-}
import           System.Log.Logger ( updateGlobalLogger
                                   , rootLoggerName
                                   , setLevel
                                   , debugM
                                   , Priority(..)
                                   )
import           Control.Monad.IO.Class (liftIO)
import           Control.Monad          (msum)
import           Data.List         (intercalate)
import           Data.Text         (Text, pack, unpack)
import           Happstack.Server  
import           Database.SQLite.Simple
import           Database.SQLite.Simple.FromRow
import           Data.Aeson
import           GHC.Generics                  (Generic)
import qualified Data.ByteString.Lazy.Char8 as BC

data WeatherField = WeatherField {date :: Text, temperature :: Float}
                    deriving (Generic, Show)

instance FromRow WeatherField where -- ^ Marshal data from DB to our ADT
  fromRow = WeatherField <$> field <*> field 

instance ToRow WeatherField where -- ^ Marshal data from our ADT to the DB
  toRow (WeatherField theDate temp) = toRow (theDate, temp)

instance ToJSON   WeatherField -- ^ Marshal data from our ADT to JSON
instance FromJSON WeatherField -- ^ Marshal data from JSON to our ADT

{-| Handle requests for a single date. -}
dayHandler :: Text -> Connection -> ServerPart Response
dayHandler d conn = do
  r <- liftIO (queryNamed conn "SELECT the_date, temperature \
                               \ FROM  weather \
                               \ WHERE the_date = :dt"
                               [":dt" := d] :: IO [WeatherField])
  liftIO $ debugM "Date Query" $ listToOutput r -- ^ NB example of how to output debug messages
  case r of
    [] -> notFoundHandler
    _  -> ok $ toResponse $ listToOutput r

{-| Handle PUT requests for date/temperature pairs. -}
--Not Working
dayPutHandler :: Text -> Text -> Connection -> ServerPart Response
dayPutHandler d t conn = do
  r <- liftIO (queryNamed conn "SELECT the_date, temperature \
                               \ FROM  weather \
                               \ WHERE the_date = :dt"
                               [":dt" := d] :: IO [WeatherField])
  liftIO $ debugM "Date PUT request" $ listToOutput r
  case r of
    [] -> insertHandler d t conn
    _  -> updateHandler d t conn

{-| Insert a new date/temperature pair. -}
insertHandler :: Text -> Text -> Connection -> ServerPart Response
insertHandler d t conn = do
  let t' = (read $ unpack t)::Float
  liftIO (execute conn "INSERT INTO weather (the_date, temperature) \
                       \ VALUES (?,?)" (WeatherField d t'))
  ok emptyJSONResponse

{-| Update a date/temperature pair. -}
updateHandler :: Text -> Text -> Connection -> ServerPart Response
updateHandler d t conn = do
  liftIO (executeNamed conn "UPDATE weather \
                          \ SET temperature = :dt1 \
                          \ WHERE the_date = :dt2"
                          [":dt1" := t, ":dt2" := d])
  ok emptyJSONResponse

{-| Return 404 Not Found and an empty JSON object -}
notFoundHandler :: ServerPart Response
notFoundHandler = notFound emptyJSONResponse

{-| An empty JSON object -}
--emptyJSONObject :: ServerPart Response
emptyJSONResponse = toResponse $ pack "[]"

{-| Turn a list of WeatherFields into a JSON object. -}
listToOutput :: ToJSON a => [a] -> String
listToOutput xs = "[" ++ intercalate "," (map (BC.unpack . encode) xs) ++ "]"

{-| Handle requests for dates in between date 1 and date 2.-}
rangeHandler :: Text -> Text -> Connection -> ServerPart Response
rangeHandler d1 d2 conn = do
  r <- liftIO (queryNamed conn "SELECT the_date, temperature \
                               \ FROM  weather \
                               \ WHERE the_date BETWEEN :dt1 AND :dt2"
                               [":dt1" := d1,":dt2" := d2] :: IO [WeatherField])
  liftIO $ debugM "Range Query" $ listToOutput r
  case r of
    [] -> notFoundHandler
    _  -> ok $ toResponse $ listToOutput r

{-| Handle requests for dates in between date 1 and date 2 with max temperature.-}
maxHandler :: Text -> Text -> Connection -> ServerPart Response
maxHandler d1 d2 conn = do
  r <- liftIO (queryNamed conn "SELECT the_date, MAX(temperature) \
                               \ FROM  weather \
                               \ WHERE the_date BETWEEN :dt1 AND :dt2"
                               [":dt1" := d1,":dt2" := d2] :: IO [WeatherField])
  liftIO $ debugM "Max Query" $ listToOutput r
  case r of
    [] -> notFoundHandler
    _  -> ok $ toResponse $ listToOutput r

{-| Handle requests for dates that has a higher temperature than t. -}
aboveTHandler :: Text -> Connection -> ServerPart Response
aboveTHandler t conn = do
  r <- liftIO (queryNamed conn "SELECT the_date, temperature \
                               \ FROM  weather \
                               \ WHERE temperature >= :dt"
                               [":dt" := t] :: IO [WeatherField])
  liftIO $ debugM "Above Query" $ listToOutput r
  case r of
    [] -> notFoundHandler
    _  -> ok $ toResponse $ listToOutput r

{-| Return 405 Method Not Allowed and an empty JSON object -}
methodNotAllowedHandler :: ServerPart Response
methodNotAllowedHandler = resp 405 emptyJSONResponse

src/WeatherService/Service.hs
{-# LANGUAGE OverloadedStrings, FlexibleContexts, TypeFamilies
  , QuasiQuotes, TemplateHaskell, DeriveGeneric #-}
module WeatherService.Service (WeatherField(..)
                                , site
                                , homePage
                                , dayHandler
                                , putHandler
                                , rangeHandler
                                , maxHandler
                                , aboveTHandler) where
{-| Semester 2 assignment for CI285, University of Brighton
    Jim Burton <j.burton@brighton.ac.uk>
    Modifier: Chris Tran - 15800120
-}
import           System.Log.Logger          (debugM)
import           Control.Monad.IO.Class     (liftIO)
import           Control.Monad              (msum)
import           Data.List                  (intercalate)
import           Data.Text                  (Text, pack, unpack)

import           Control.Applicative

import           Happstack.Server
import           Web.Routes
import           Web.Routes.Happstack
import           Web.Routes.TH

import           Database.SQLite.Simple
import           Database.SQLite.Simple.FromRow

import           Data.Aeson
import           GHC.Generics               (Generic)
import qualified Data.ByteString.Lazy.Char8 as BC

data WeatherField = WeatherField {date :: Text, temperature :: Float}
                    deriving (Generic, Show)

instance FromRow WeatherField where -- ^ Marshal data from DB to our ADT
  fromRow = WeatherField <$> field <*> field

instance ToRow WeatherField where -- ^ Marshal data from our ADT to the DB
  toRow (WeatherField theDate temp) = toRow (theDate, temp)

instance ToJSON   WeatherField -- ^ Marshal data from our ADT to JSON
instance FromJSON WeatherField -- ^ Marshal data from JSON to our ADT

{-| Site Map -}
data Sitemap
  = Weather           --weather/                      (home page)
  | Date Text         --weather/date/YYYY-mm-dd
  | Put Text Text     --weather/put/date/temperature
  | Range Text Text   --weather/range/date1/date2
  | Max Text Text     --weather/max/date1/date2
  | Above Text        --weather/above/temperature

$(derivePathInfo ''Sitemap)

{-| Site Route -}
route :: Connection -> Sitemap -> RouteT Sitemap (ServerPartT IO) Response
route conn url =
  case url of
    Weather        -> (method [GET, POST] >> homePage)
                      <|> methodNotAllowedHandler
    (Date d)       -> (method [GET, POST] >> dayHandler d conn)
                      <|> methodNotAllowedHandler
    (Put d t)      -> (method PUT >> putHandler d t conn)
                      <|> methodNotAllowedHandler
    (Range d1 d2)  -> (method GET >> rangeHandler d1 d2 conn)
                      <|> methodNotAllowedHandler
    (Max d1 d2)    -> (method GET >> maxHandler d1 d2 conn)
                      <|> methodNotAllowedHandler
    (Above t)      -> (method GET >> aboveTHandler t conn)
                      <|> methodNotAllowedHandler

{-| Home Page -}
homePage :: RouteT Sitemap (ServerPartT IO) Response
homePage = do
  setHeaderM "Content-Type" "text/plain"
  ok $ toResponse $
     BC.pack "Home Page\n\
             \Haskell Weather Service by Chris Tran\n\
             \\n\
             \*Note: d = date (YYYY-mm-dd)\n\
             \       t = temperature\n\
             \\n\
             \To check temperature of a single day:\n\
             \  /weather/date/d\n\
             \\n\
             \To insert or update new date/temperature pair:\n\
             \In terminal:\n\
             \  $ curl -X PUT http://localhost:8000/weather/put/d/t\n\
             \\n\
             \To check temperature in between a range:\n\
             \  /weather/range/d1/d2\n\
             \\n\
             \To find day(s) that has max temperature in between a range:\n\
             \  /weather/max/d1/d2\n\
             \\n\
             \To find day(s) that has higher temperature than t:\n\
             \  /weather/above/t\n"

{-| Handle requests for a single date. -}
dayHandler :: Text -> Connection -> RouteT Sitemap (ServerPartT IO) Response
dayHandler d conn = do
  r <- liftIO (queryNamed conn "SELECT the_date, temperature \
                               \ FROM  weather \
                               \ WHERE the_date = :dt"
                               [":dt" := d] :: IO [WeatherField])
  liftIO $ debugM "Date Query" $ listToOutput r
  case r of
    [] -> notFoundHandler
    _  -> ok $ toResponse $ listToOutput r

{-| Handle PUT requests for date/temperature pairs. -}
putHandler :: Text -> Text -> Connection -> RouteT Sitemap (ServerPartT IO) Response
putHandler d t conn = do
  r <- liftIO (queryNamed conn "SELECT the_date, temperature \
                               \ FROM  weather \
                               \ WHERE the_date = :dt"
                               [":dt" := d] :: IO [WeatherField])
  liftIO $ debugM "Put request" $ listToOutput r
  case r of
    [] -> insertHandler d t conn
    _  -> updateHandler d t conn

{-| Insert a new date/temperature pair. -}
insertHandler :: Text -> Text -> Connection -> RouteT Sitemap (ServerPartT IO) Response
insertHandler d t conn = do
  let t' = (read $ unpack t)::Float
  liftIO (execute conn "INSERT INTO weather (the_date, temperature) \
                       \ VALUES (?,?)" (WeatherField d t'))
  ok emptyJSONResponse

{-| Update a date/temperature pair. -}
updateHandler :: Text -> Text -> Connection -> RouteT Sitemap (ServerPartT IO) Response
updateHandler d t conn = do
  liftIO (executeNamed conn "UPDATE weather \
                          \ SET temperature = :dt1 \
                          \ WHERE the_date = :dt2"
                          [":dt1" := t, ":dt2" := d])
  ok emptyJSONResponse

{-| Return 404 Not Found and an empty JSON object -}
notFoundHandler :: RouteT Sitemap (ServerPartT IO) Response
notFoundHandler = notFound emptyJSONResponse

{-| An empty JSON object -}
emptyJSONResponse = toResponse $ BC.pack "[]"

{-| Turn a list of WeatherFields into a JSON object. -}
listToOutput :: ToJSON a => [a] -> String
listToOutput xs = "[" ++ intercalate "," (map (BC.unpack . encode) xs) ++ "]"

{-| Handle requests for dates in between date 1 and date 2.-}
rangeHandler :: Text -> Text -> Connection -> RouteT Sitemap (ServerPartT IO) Response
rangeHandler d1 d2 conn = do
  r <- liftIO (queryNamed conn "SELECT the_date, temperature \
                               \ FROM  weather \
                               \ WHERE the_date BETWEEN :dt1 AND :dt2"
                               [":dt1" := d1,":dt2" := d2] :: IO [WeatherField])
  liftIO $ debugM "Range Query" $ listToOutput r
  case r of
    [] -> notFoundHandler
    _  -> ok $ toResponse $ listToOutput r

{-| Handle requests for dates in between date 1 and date 2 with max temperature.-}
maxHandler :: Text -> Text -> Connection -> RouteT Sitemap (ServerPartT IO) Response
maxHandler d1 d2 conn = do
  r <- liftIO (queryNamed conn "SELECT the_date, MAX(temperature) \
                               \ FROM  weather \
                               \ WHERE the_date BETWEEN :dt1 AND :dt2"
                               [":dt1" := d1,":dt2" := d2] :: IO [WeatherField])
  liftIO $ debugM "Max Query" $ listToOutput r
  case r of
    [] -> notFoundHandler
    _  -> ok $ toResponse $ listToOutput r

{-| Handle requests for dates that has a higher temperature than t. -}
aboveTHandler :: Text -> Connection -> RouteT Sitemap (ServerPartT IO) Response
aboveTHandler t conn = do
  r <- liftIO (queryNamed conn "SELECT the_date, temperature \
                               \ FROM  weather \
                               \ WHERE temperature >= :dt"
                               [":dt" := t] :: IO [WeatherField])
  liftIO $ debugM "Above Temperature Query" $ listToOutput r
  case r of
    [] -> notFoundHandler
    _  -> ok $ toResponse $ listToOutput r

{-| Return 405 Method Not Allowed and an empty JSON object -}
methodNotAllowedHandler :: RouteT Sitemap (ServerPartT IO) Response
methodNotAllowedHandler = resp 405 emptyJSONResponse

{-| Site with Weather set as Home Page -}
site :: Connection -> Site Sitemap (ServerPartT IO Response)
site conn =
            setDefault Weather $ mkSitePI (runRouteT (route conn))
